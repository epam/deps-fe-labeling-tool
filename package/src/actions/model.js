
import { batch } from 'react-redux'
import { createAction } from 'redux-actions'
import { LabelType, LABEL_TYPE_NAME } from '@/models/Label'
import { Markup } from '@/models/Markup'
import { markupSelector } from '@/selectors/markup'
import { fieldsSelector, fieldsToDeleteSelector } from '@/selectors/model'
import {
  updateLabelsWithSettings,
  updateTablesWithSettings
} from './markup'

const FEATURE_NAME = 'MODEL'

const storeFields = createAction(
  `${FEATURE_NAME}/STORE`
)

const storeInitialFields = createAction(
  `${FEATURE_NAME}/STORE_INITIAL_FIELDS`
)

const updateInitialFields = createAction(
  `${FEATURE_NAME}/UPDATE_INITIAL_FIELDS`
)

const enableDeleteFieldsMode = createAction(
  `${FEATURE_NAME}/ENABLE_DELETE_FIELD_MODE`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

const updateFieldsToDelete = createAction(
  `${FEATURE_NAME}/UPDATE_FIELDS_TO_DELETE`
)

const cancelDeleteFieldsMode = createAction(
  `${FEATURE_NAME}/CANCEL_DELETE_FIELDS_MODE`
)

const confirmFieldsDeletion = () => (dispatch, getState) => {
  const state = getState()
  const markup = markupSelector(state)
  const fields = fieldsSelector(state)
  const fieldsToDelete = fieldsToDeleteSelector(state)
  const newFields = fields.filter((field) => fieldsToDelete.every((f) => f.code !== field.code))
  const assignedObjects = markup
    ? Markup.getAllObjects(markup).filter((markupObject) => fieldsToDelete.find((field) => field.code === markupObject.fieldCode))
    : []

  batch(() => {
    assignedObjects.forEach((markupObject) => {
      if (markupObject.typeName === LABEL_TYPE_NAME) {
        dispatch(updateLabelsWithSettings({
          ...markupObject,
          fieldCode: '',
          type: LabelType.UNASSIGNED,
          index: undefined
        }))
      } else {
        dispatch(updateTablesWithSettings({
          ...markupObject,
          fieldCode: '',
          index: undefined
        }))
      }
    })
    dispatch(storeFields(newFields))
    dispatch(cancelDeleteFieldsMode())
  })
}

export {
  enableDeleteFieldsMode,
  resetDefault,
  storeFields,
  storeInitialFields,
  updateFieldsToDelete,
  updateInitialFields,
  confirmFieldsDeletion,
  cancelDeleteFieldsMode
}
