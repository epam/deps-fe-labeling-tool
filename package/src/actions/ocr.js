import { createAction } from 'redux-actions'

const FEATURE_NAME = 'OCR'

const storeOcr = createAction(`${FEATURE_NAME}/STORE_OCR`)

const setPrimaryEngine = createAction(`${FEATURE_NAME}/SET_PRIMARY_ENGINE`)

const setPrimaryLanguage = createAction(`${FEATURE_NAME}/SET_PRIMARY_LANGUAGE`)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  setPrimaryEngine,
  setPrimaryLanguage,
  storeOcr
}
