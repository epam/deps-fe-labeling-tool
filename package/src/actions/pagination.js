import { createAction } from 'redux-actions'

const FEATURE_NAME = 'PAGINATION'

const openPage = createAction(`${FEATURE_NAME}/OPEN_PAGE`)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  openPage
}
