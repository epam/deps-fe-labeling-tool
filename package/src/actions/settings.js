import { createAction } from 'redux-actions'

const FEATURE_NAME = 'SETTINGS'

const storeSettings = createAction(`${FEATURE_NAME}/STORE`)

const resetDefault = createAction(`${FEATURE_NAME}/RESET_DEFAULT`)

export {
  resetDefault,
  storeSettings
}
