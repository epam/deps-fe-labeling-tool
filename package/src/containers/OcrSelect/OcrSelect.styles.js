import styled from 'styled-components'
import { Select } from '@/components/Select'
import { COLORS } from '@/theme/theme.default'

const StyledSelect = styled(Select)`
  font-size: 1.6rem;
  line-height: 2.2rem;
  color: ${COLORS.PRIMARY_4_TRANSPARENT};
`

export {
  StyledSelect as Select
}
