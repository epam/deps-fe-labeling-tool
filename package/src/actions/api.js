import { createAction } from 'redux-actions'
import { rotationAnglesSelector } from '@/selectors/canvas'
import { pageImageUrlSelector } from '@/selectors/document'
import { markupSelector } from '@/selectors/markup'
import { fieldsSelector } from '@/selectors/model'
import { primaryLanguageSelector } from '@/selectors/ocr'
import { getApi } from '@/services/api'

const FEATURE_NAME = 'API'

const requestAttempt = createAction(
  `${FEATURE_NAME}/REQUEST_ATTEMPT`
)

const requestSuccess = createAction(
  `${FEATURE_NAME}/REQUEST_SUCCESS`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

const requestFailure = createAction(
  `${FEATURE_NAME}/REQUEST_FAILURE`,
  (requestId, error) => ({
    requestId,
    error
  })
)

export const createRequestAction = (requestId, actionCreator) => {
  const requestActionCreator = (...args) => async (dispatch, getState) => {
    try {
      dispatch(requestAttempt(requestId))
      const result = await actionCreator(...args)(dispatch, getState)
      dispatch(requestSuccess(requestId))
      return result
    } catch (error) {
      dispatch(requestFailure(requestId, error.message))
      throw error
    }
  }
  requestActionCreator.toString = () => requestId
  return requestActionCreator
}

const save = createRequestAction(
  'save',
  () => (dispatch, getState) => {
    const state = getState()
    const markup = markupSelector(state)
    const rotationAngles = rotationAnglesSelector(state)
    const language = primaryLanguageSelector(state)
    return getApi().save(markup, rotationAngles, language)
  }
)

const saveMarkup = createRequestAction(
  'saveMarkup',
  () => (dispatch, getState) => {
    const state = getState()
    const markup = markupSelector(state)
    const rotationAngles = rotationAnglesSelector(state)
    const language = primaryLanguageSelector(state)
    const fields = fieldsSelector(state)
    return getApi().saveMarkup(markup, rotationAngles, language, fields)
  }
)

const ocrText = createRequestAction(
  'ocrText',
  (engine, coordinates) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)
    const language = primaryLanguageSelector(state)
    return getApi().ocrText(
      engine,
      pageImageUrl,
      coordinates,
      language
    )
  }
)

const omrArea = createRequestAction(
  'omrArea',
  (coordinates) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)

    return getApi().omrArea(
      pageImageUrl,
      coordinates
    )
  }
)

const detectTables = createRequestAction(
  'detectTables',
  (selectedAreaCoords) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)
    return getApi().detectTables(
      pageImageUrl,
      selectedAreaCoords
    )
  }
)

const ocrTable = createRequestAction(
  'ocrTable',
  (engine, table) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)
    const language = primaryLanguageSelector(state)
    return getApi().ocrTable(
      engine,
      pageImageUrl,
      table,
      language
    )
  }
)

export {
  detectTables,
  ocrTable,
  ocrText,
  omrArea,
  resetDefault,
  requestAttempt,
  requestSuccess,
  requestFailure,
  save,
  saveMarkup
}
