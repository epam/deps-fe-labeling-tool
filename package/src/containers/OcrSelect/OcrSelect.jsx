import React, { useCallback, useMemo, useEffect } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { setPrimaryEngine } from '@/actions/ocr'
import { HotKeyEvent } from '@/constants/hotKeys'
import { withHotKeys } from '@/hocs/withHotKeys'
import { ocrEngineShape } from '@/models/OcrEngine'
import { Option } from '@/models/Option'
import {
  primaryEngineSelector,
  ocrEnginesSelector
} from '@/selectors/ocr'
import { Select } from './OcrSelect.styles'

const OcrSelect = ({
  setPrimaryEngine,
  primaryEngine,
  registerHandlers,
  ocr,
  ocrEngines
}) => {
  const onSelectItemChange = useCallback((value) => {
    const engine = ocrEngines.find((engine) => engine.code === value)
    setPrimaryEngine(engine.code)
  }, [ocrEngines, setPrimaryEngine])

  const hotKeyHandlers = useMemo(() => ({
    [HotKeyEvent.RECOGNIZE]: ocr
  }), [ocr])

  useEffect(() => {
    registerHandlers(hotKeyHandlers)
  }, [hotKeyHandlers, registerHandlers])

  const generateOptions = () => (
    ocrEngines.map((engine) => new Option(engine.code, engine.name))
  )

  return (
    <Select
      onChange={onSelectItemChange}
      defaultValue={primaryEngine}
      options={generateOptions()}
      getPopupContainer={() => document.body}
    />
  )
}

OcrSelect.propTypes = {
  ocrEngines: PropTypes.arrayOf(ocrEngineShape).isRequired,
  primaryEngine: PropTypes.string.isRequired,
  setPrimaryEngine: PropTypes.func.isRequired,
  registerHandlers: PropTypes.func.isRequired,
  ocr: PropTypes.func.isRequired
}

const OcrDropdownButtonWithHotKeys = withHotKeys(OcrSelect)

const mapStateToProps = (state) => ({
  primaryEngine: primaryEngineSelector(state),
  ocrEngines: ocrEnginesSelector(state)
})

const OcrDropdownButtonContainer = connect(mapStateToProps, { setPrimaryEngine })(OcrDropdownButtonWithHotKeys)

export {
  OcrDropdownButtonContainer as OcrSelect
}
