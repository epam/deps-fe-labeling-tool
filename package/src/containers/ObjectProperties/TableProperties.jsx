import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { ocrTable } from '@/actions/api'
import { updateTables, updateAllTables } from '@/actions/markup'
import { setActiveSidebar } from '@/actions/ui'
import { ButtonType } from '@/components/Button'
import { TableIcon } from '@/components/Icons/TableIcon'
import { ObjectTitle } from '@/components/ObjectTitle'
import { ObjectCoordinates } from '@/containers/ObjectCoordinates'
import { ObjectNameAndIndex } from '@/containers/ObjectNameAndIndex'
import { OcrSelect } from '@/containers/OcrSelect'
import { FieldType } from '@/enums/FieldType'
import { SidebarContent } from '@/enums/SidebarContent'
import { fieldShape } from '@/models/Field'
import { ocrEngineShape } from '@/models/OcrEngine'
import * as Table from '@/models/Table'
import { fieldsSelector } from '@/selectors/model'
import { primaryEngineSelector, ocrEnginesSelector } from '@/selectors/ocr'
import { currentPageSelector } from '@/selectors/pagination'
import { settingsSelector } from '@/selectors/settings'
import { getApi } from '@/services/api'
import { Grid, Divider, Button } from './ObjectProperties.styles'
import {
  TableDataName,
  StyledOpenButton,
  TableData
} from './TableProperties.styles'

const RESOURCE_OCR = 'OCR'
const RESOURCE_CONTENT = 'CONTENT'
const RESOURCE_COORDINATES = 'COORDINATES'
const RESOURCE_EXTRACT_DATA = 'Extract data'
const RESOURCE_TABLE_DATA = 'Table Data'
const RESOURCE_OPEN = 'Open'

const { tableShape } = Table

class TableProperties extends PureComponent {
  static propTypes = {
    currentPage: PropTypes.number.isRequired,
    fields: PropTypes.arrayOf(fieldShape).isRequired,
    setActiveSidebar: PropTypes.func.isRequired,
    ocrTable: PropTypes.func.isRequired,
    ocrEngines: PropTypes.arrayOf(ocrEngineShape),
    primaryEngine: PropTypes.string.isRequired,
    table: tableShape.isRequired,
    updateAllTables: PropTypes.func.isRequired,
    updateTables: PropTypes.func.isRequired
  }

  octTable = async (selectedEngine) => {
    const newTable = await this.props.ocrTable(selectedEngine, this.props.table)
    this.props.updateTables(this.props.currentPage, [
      newTable
    ])
  }

  onColumnsChange = (value) => {
    const { rows } = Table.getDimensions(this.props.table)
    this.props.updateTables(this.props.currentPage, [
      Table.divide(this.props.table, rows, value)
    ])
  }

  onRowsChange = (value) => {
    const { columns } = Table.getDimensions(this.props.table)
    this.props.updateTables(this.props.currentPage, [
      Table.divide(this.props.table, value, columns)
    ])
  }

  getTableFields = () => this.props.fields.filter((f) => {
    if (f.fieldType === FieldType.LIST) {
      return f.fieldMeta.baseType === FieldType.TABLE
    }
    return f.fieldType === FieldType.TABLE
  })

  onOpenTableDataClick = () => {
    this.props.setActiveSidebar(SidebarContent.TABLE_DATA)
  }

  isTableOCRAvailable = () => !!this.props.ocrEngines.length && !!getApi().ocrTable

  render = () => {
    return (
      <Grid>
        <ObjectTitle
          title={RESOURCE_CONTENT}
        />
        <ObjectNameAndIndex
          object={this.props.table}
          onChange={this.props.updateAllTables}
        />
        {
          this.isTableOCRAvailable() && (
            <Button
              onClick={() => this.octTable(this.props.primaryEngine)}
              type={ButtonType.PRIMARY}
            >
              {RESOURCE_EXTRACT_DATA}
            </Button>
          )
        }
        <TableData>
          <TableDataName>
            <TableIcon />
            {RESOURCE_TABLE_DATA}
          </TableDataName>
          <StyledOpenButton onClick={this.onOpenTableDataClick}>
            {RESOURCE_OPEN}
          </StyledOpenButton>
        </TableData>
        <Divider />
        <ObjectTitle
          title={RESOURCE_COORDINATES}
        />
        <ObjectCoordinates />
        {
          this.isTableOCRAvailable() && (
            <>
              <Divider />
              <ObjectTitle
                title={RESOURCE_OCR}
              />
              <OcrSelect
                ocr={this.octTable}
              />
            </>
          )
        }
      </Grid>
    )
  }
}

const mapStateToProps = (state) => ({
  currentPage: currentPageSelector(state),
  fields: fieldsSelector(state),
  ocrEngines: ocrEnginesSelector(state),
  settings: settingsSelector(state),
  primaryEngine: primaryEngineSelector(state)
})

const mapDispatchToProps = {
  ocrTable,
  setActiveSidebar,
  updateTables,
  updateAllTables
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(TableProperties)

export {
  ConnectedComponent as TableProperties
}
