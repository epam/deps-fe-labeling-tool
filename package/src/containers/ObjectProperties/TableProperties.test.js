import { mockApiActions } from '@/mocks/actions/api'
import { mockMarkupActions } from '@/mocks/actions/markup'
import { mockComponent } from '@/mocks/mockComponent'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockUuid } from '@/mocks/mockUuid'
import { mockModelSelectors } from '@/mocks/selectors/model'
import { mockOcrSelectors } from '@/mocks/selectors/ocr'
import { mockPaginationSelectors } from '@/mocks/selectors/pagination'
import { mockSettingsSelectors } from '@/mocks/selectors/settings'
import React from 'react'
import { shallow } from 'enzyme'
import { ocrTable } from '@/actions/api'
import { updateAllTables, updateTables } from '@/actions/markup'
import { OcrSelect } from '@/containers/OcrSelect'
import { Table } from '@/models/Table'
import { TableProperties } from './TableProperties'

jest.mock('react-redux', () => mockReactRedux)

jest.mock('@/selectors/pagination', () => mockPaginationSelectors)
jest.mock('@/selectors/model', () => mockModelSelectors)
jest.mock('@/selectors/ocr', () => mockOcrSelectors)
jest.mock('@/selectors/settings', () => mockSettingsSelectors)

jest.mock('@/actions/markup', () => mockMarkupActions)
jest.mock('@/actions/api', () => mockApiActions)

jest.mock('@/hocs/debounce', () => ({
  debounce: () => (Component) => Component
}))

jest.mock('@/containers/ObjectNameAndIndex', () => mockComponent('ObjectNameAndIndex'))
jest.mock('@/containers/OcrSelect', () => mockComponent('OcrSelect'))
jest.mock('@/containers/ObjectCoordinates', () => mockComponent('ObjectCoordinates'))

const mockContentTable = 'mockContent'
const mockOcrTable = jest.fn(() => Promise.resolve(mockContentTable))
jest.mock('@/services/api', () => ({
  getApi: jest.fn(() => ({
    ocrTable: mockOcrTable
  }))
}))

jest.mock('uuid', () => mockUuid)

const { WrappedComponent, mapStateToProps, mapDispatchToProps } = TableProperties

describe('Container: TableProperties', () => {
  describe('mapStateToProps', () => {
    it('should pass expected state as props to WrappedComponent', () => {
      const { props } = mapStateToProps()
      expect(props).toMatchSnapshot()
    })
  })

  describe('mapDispatchToProps', () => {
    const { props } = mapDispatchToProps()

    it('should pass ocrTable action as prop to WrappedComponent', async () => {
      await props.ocrTable()
      expect(ocrTable).toHaveBeenCalledTimes(1)
    })

    it('should pass updateTables action as prop to WrappedComponent', () => {
      props.updateTables([])
      expect(updateTables).nthCalledWith(1, [])
    })

    it('should pass updateAllTables action as prop to WrappedComponent', () => {
      const updateAllTablesPayload = {
        newTable: new Table([0.1, 0.33], [0.32, 0.55])
      }
      props.updateAllTables(updateAllTablesPayload)
      expect(updateAllTables).nthCalledWith(1, updateAllTablesPayload)
    })
  })

  describe('component', () => {
    let wrapper, defaultProps
    const testTable = new Table([0.21, 0.33], [0.1, 0.51])

    beforeEach(() => {
      defaultProps = {
        ...mapStateToProps().props,
        ...mapDispatchToProps().props,
        table: testTable,
        primaryEngine: 'ocrTable'
      }

      wrapper = shallow(<WrappedComponent {...defaultProps} />)
    })

    it('should render layout correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call prop updateTables with correct args when calling OcrDropdownButton ocr prop (ocrTable)', async () => {
      const OcrDropdownButtonWrapper = wrapper.find(OcrSelect)
      await OcrDropdownButtonWrapper.props().ocr('mockEngine')
      expect(defaultProps.updateTables).nthCalledWith(1, defaultProps.currentPage, [{ type: 'MOCK_OCR_TABLE' }])
    })
  })
})
