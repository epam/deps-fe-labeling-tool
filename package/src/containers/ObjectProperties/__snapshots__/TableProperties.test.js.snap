// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: TableProperties component should render layout correctly 1`] = `
<ObjectPropertiesstyles__Grid>
  <ObjectTitle
    title="CONTENT"
  />
  <ObjectNameAndIndex
    object={
      Table {
        "fieldCode": "",
        "index": undefined,
        "merges": Array [],
        "meta": Object {},
        "typeName": "table",
        "uid": "6",
        "values": Array [],
        "xGuidelines": Array [
          0.21,
          0.33,
        ],
        "yGuidelines": Array [
          0.1,
          0.51,
        ],
      }
    }
    onChange={[MockFunction]}
  />
  <ObjectPropertiesstyles__StyledButton
    onClick={[Function]}
    type="primary"
  >
    Extract data
  </ObjectPropertiesstyles__StyledButton>
  <TablePropertiesstyles__TableData>
    <TablePropertiesstyles__TableDataName>
      <TableIcon />
      Table Data
    </TablePropertiesstyles__TableDataName>
    <TablePropertiesstyles__StyledOpenButton
      onClick={[Function]}
    >
      Open
    </TablePropertiesstyles__StyledOpenButton>
  </TablePropertiesstyles__TableData>
  <ObjectPropertiesstyles__StyledDivider />
  <ObjectTitle
    title="COORDINATES"
  />
  <ObjectCoordinates />
  <ObjectPropertiesstyles__StyledDivider />
  <ObjectTitle
    title="OCR"
  />
  <OcrSelect
    ocr={[Function]}
  />
</ObjectPropertiesstyles__Grid>
`;

exports[`Container: TableProperties mapStateToProps should pass expected state as props to WrappedComponent 1`] = `
Object {
  "currentPage": 1,
  "fields": Array [
    Field {
      "code": "123",
      "fieldType": "string",
      "name": "TestField",
      "required": false,
    },
    Field {
      "code": "enum",
      "fieldMeta": Object {
        "options": Array [
          "test",
          "test1",
        ],
      },
      "fieldType": "enum",
      "name": "TestField",
      "required": false,
    },
    Field {
      "code": "enumList",
      "fieldMeta": ListFieldMeta {
        "baseType": "enum",
        "baseTypeMeta": EnumFieldMeta {
          "options": Array [
            "test",
            "test1",
          ],
        },
      },
      "fieldType": "list",
      "name": "EnumerationList",
      "required": false,
    },
  ],
  "ocrEngines": Array [
    OcrEngine {
      "code": "ocrCode",
      "name": "ocr Name",
    },
  ],
  "primaryEngine": "ocrCode",
  "settings": Settings {
    "features": Array [
      SubSettings {
        "code": "autoSave",
        "data": Object {
          "interval": 180000,
        },
      },
      "export",
      "import",
      SubSettings {
        "code": "paging",
        "data": Object {
          "initialPage": 1,
        },
      },
      "language",
      "rotation",
      SubSettings {
        "code": "specialSymbols",
        "data": Array [
          "&",
          "$",
          "£",
          "¥",
          "©",
          "®",
          "¢",
          "%",
          "*",
          "@",
          "µ",
          "±",
          "´",
          "²",
          "~",
          "½",
          "«",
          "»",
          "°",
          "–",
          "—",
          "“",
          "”",
        ],
      },
      "manageFields",
      "multiAssignLabels",
      "multiAssignTables",
    ],
    "mode": "default",
    "panels": Array [
      "TOOLBAR",
      "MARKUP_SIDEBAR",
      "LEFT_SIDEBAR",
    ],
    "tools": Array [
      "detectTables",
      "label",
      "table",
      "merge",
      "split",
      "pointer",
      "grabbing",
    ],
  },
}
`;
